### PR TITLE
Dra 640 login keycloak oath

### PR DESCRIPTION
**Pull request by thomasegense**
See internal jira DRA-640 for information

### CODE REVIEW

File: src/main/java/dk/kb/license/util/OauthUtil.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+        byte[] keyBytes = Base64.getDecoder().decode(publicKeyBase64.getBytes(StandardCharsets.UTF_8));
+        X509EncodedKeySpec specPublic = new X509EncodedKeySpec(keyBytes);
+        PublicKey fileGeneratedPublicKey = keyFactory.generatePublic(specPublic);
+        RSAPublicKey rsaPub  = (RSAPublicKey)(fileGeneratedPublicKey);
+        return rsaPub;       
+    }
+    
+    private static JWTVerifier buildJWTVerifier(String rsaPublicKeyBase64) throws Exception {
+        var algo = Algorithm.RSA256(getRSAPublicKey(rsaPublicKeyBase64), null);
```
Comment (VictorHarbo) [inline src/main/java/dk/kb/license/util/OauthUtil.java:233]: Should the type be Algorithm instead of var?

**Review by VictorHarbo — CHANGES_REQUESTED** [review]
A few comments here and there. @davidgrove73 should probably give a more thorough review on the security of the implementation.

File: conf/ds-license-environment.yaml.SAMPLE
[the remark this answers is not in the document]
Comment (thomasegense) [reply conf/ds-license-environment.yaml.SAMPLE:1]: added.

File: src/main/java/dk/kb/license/servlets/LoginCallbackServlet.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        catch(Exception e) {
+          log.error("Error validating redirect code against keycloak",e);
+          throw new ServletException(e);
+          
+        }        
+        log.info("Login from KeyCloak success for user:"+combinedUserName);
+        request.getSession().setAttribute("oauth_user", combinedUserName); //Save on session           
+        response.sendRedirect(request.getContextPath() + "/configuration.jsp"); //Default landing page
+        return;
```
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/servlets/LoginCallbackServlet.java:61]: Removed return

File: src/main/java/dk/kb/license/servlets/LoginCallbackServlet.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     * <p>
+     * After login at KeyCloak users are redirected to this servlet. <br>
+     * Login URL at KeyCloak: https://<keycloak_server>realms/master/protocol/openid-connect/auth?response_type=code&client_id=kb-ad&redirect_uri=http://<redirect-server>/ds-license/loginCallbackServlet<br>
+     * After successful login, users are redirected back to this servlet.<br>
+     * Example redirect url: http://<redirect-server>/ds-license/loginCallbackServlet?session_state=96bff8da-63d4-4f61-850d-23317ece7f90&iss=https%3A%2F%2F<keycloak_server>%2Frealms%2Fmaster&code=b6374b3d-b182-4e9f-acd1-20cf00a2c0d2.96bff8da-63d4-4f61-850d-23317ece7f90.34931351-1c06-48e8-a120-36ae04e1bff1<b>
+     * </p>
+     * The url parameter 'code' and keycloak secret (from property file) are used to fetch user data from keycloak and validate redirect came from keycloak.<br>
+     * The response from keycloak is JWT encoded. The value user name and user email are extracted from the JWT and set on session to mark user is logged in.<br>
+     *  
+     */
```
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/servlets/LoginCallbackServlet.java:37]: Not much to describe. Is called by Tomcat automatic. All servlet doGet/doPost has this API.

**Review by thomasegense — COMMENTED** [review]
Added comments.

**Review by davidgrove73 — APPROVED** [review]

**Pull request merged**